pyop2/codegen/c/solve.c

```diff
@@ -10,17 +11,36 @@ static PetscScalar work_buffer[BUF_SIZE*BUF_SIZE];
 
 static void solve(PetscScalar* __restrict__ out, const PetscScalar* __restrict__ A, const PetscScalar* __restrict__ B, PetscBLASInt N)
 {
+    PetscLogEvent  USER_EVENT_memcpy;
+    PetscLogEvent  USER_EVENT_getrf;
+    PetscLogEvent  USER_EVENT_getrs;
+    PetscClassId   classid;
+
+    PetscClassIdRegister("PyOP2SolveCallable",&classid);
+    PetscLogEventRegister("PyOP2SolveCallable_memcpy",classid,&USER_EVENT_memcpy);
+    PetscLogEventRegister("PyOP2SolveCallable_getrf",classid,&USER_EVENT_getrf);
+    PetscLogEventRegister("PyOP2SolveCallable_gerts",classid,&USER_EVENT_getrs);
```

Comment:
This allocates memory every time around, so this should be done as something like:
```c
static PetscLogEvent USER_EVENT_memcpy;
...;
static void solve(...) {
    PetscLogEventBegin(...)
}
```
Then in Python:
```python
memcpy_event = PETSc.Log.Event("some_name")
dll = somehow_load_the_dll
setattr(dll, "USER_EVENT_memcpy", memcpy_event.handle)
```
I think

Reply:
Done in https://github.com/OP2/PyOP2/pull/646/commits/f55851f4ed6dedde3bd8adeb0e63c64c6f0e7da3

Reply:
Okay so now I am not only profiling the inverse and the solve here, but also the local TSFC and Slate kernels. The PRs for that are here https://github.com/firedrakeproject/tsfc/pull/267 and here https://github.com/firedrakeproject/firedrake/pull/2347. If I want to do something morally equivalent for those events too I need a better way to do this than I have now. I guess the options are
a) we do this generic by somehow collecting all variables declares as `PetscLogEvents` and then deduce the name of the event from the naming of the event id variable
b) we keep track of event ids and names somewhere.

Reply:
If we were only dealing with loopy kernels could we not inspect `kernel.instructions` (or similar, my loopy know-how is less than perfect) to find instances where we call `PetscLogEventRegister`? I don't know how this would work with the inverse and solve callables that you have though.

Reply:
If we decide to generate the Events in python we don't need to register the events in loopy anymore. In theory, we could do what you said with `PetscLogEventBegin` but at the point where we need the event name (here https://github.com/OP2/PyOP2/blob/f55851f4ed6dedde3bd8adeb0e63c64c6f0e7da3/pyop2/compilation.py#L504) there is no loopy anymore, only C. I think what we'd need to do in order to do this in a clean way is to save the event names and event id variable names on an attribute on the `LocalKernels` and then we could iterate over that to generate the events in `compilation.py`  because the `jitmodule` is a `GlobalKernel` and carries the `LocalKernel` (or `LocalKernels`??) from what I understand. This would be a bit of work because I need to return and save the information in all the right places, so I want to be sure first that we really want to do this. Let's postpone this discussion to the next Firedrake meeting.